### ISSUE TITLE
Wrong documentation of torch.distributions.negative_binomial

### ISSUE THREAD

**Issue by bastianalt** [pytorch/pytorch#42449]
## 📚 Documentation
Strange behavior in torch.distributions.negative_binomial:
https://pytorch.org/docs/stable/distributions.html#negativebinomial

When trying out the negative binomial distribution I found that the second input parameter prob (success probability) is actually the failure probability --> not consistent with documentation. This is also not consistent with other software such as scipy.

Example:
In [1]: import torch 
   ...: from scipy.stats import nbinom 
   ...: import numpy as np 
   ...: n=2.0 
   ...: x=0.0 
   ...: p=.75 
   ...: dis=torch.distributions.NegativeBinomial(n, 1-p) 
   ...: prob_torch=dis.log_prob(torch.tensor(x)).exp() 
   ...: prob_np=nbinom.pmf(x,n,p) 
   ...: np.isclose(prob_np,prob_torch.numpy())                                  
Out[1]: True

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw @jlin27

**Comment by fritzo** [comment]
@jamestwebber do you have an opinion on the `NegativeBinomial` interface?

**Comment by jamestwebber** [comment]
> @jamestwebber do you have an opinion on the `NegativeBinomial` interface?

Not particularly, except that the documentation should match! I imagine I went through this same discovery at some point but wasn't responsible enough to raise an issue about it. Matching other implementations and resources like wikipedia would be good.

**Comment by jamestwebber** [comment]
@fritzo do you know if there's any interest in adding `GammaPoisson` to pytorch itself?

**Comment by fritzo** [comment]
@jamestwebber I don't know whether anyone else is interested in `GammaPoisson`. In Pyro we have a few conjugate-pair distributions ([conjugate.py](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/conjugate.py) and [hmm.py](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/hmm.py)) that have extra methods beyond the basic distribution interface. Also I'm holding off on porting any more Pyro distributions to PyTorch until [PyTorch allows PyTest tests #11578.

**Comment by fritzo** [comment]
cc @martinjankowiak

**Comment by SplitInfinity** [comment]
I made a PR for this (#45693) but I strongly feel that it would be better to change the code to match the docs in order to match `numpy/scipy`.

**Comment by fritzo** [comment]
@SplitInfinity I agree that it would be better to fix the code to match numpy/scipy/tensorflow. WDYT of the following interface change plan, which aims to minimize downstream breakage by including a deprecation release? (I've omitted docs changes, because they are straightforward)

1. In a first PR and released in PyTorch 1.7: widen the interface to include four optional parameters, `(success|failure)_(probs|logits)`:

<details>

```diff
- def __init__(total_count, probs=None, logits=None, validate_args=None):
+ def __init__(total_count, probs=None, logits=None, *,
+              success_probs=None, success_logits=None,
+              failure_probs=None, failure_logits=None,
+              validate_args=None):
+    # TODO remove after 1.7 release
+    if probs is not None:
+        warnings.warn("probs has been renamed to failure_probs", DeprecationWarning)
+        failure_probs = probs
+    if logits is not None:
+        warnings.warn("logits has been renamed to failure_logits", DeprecationWarning)
+        failure_logits = logits
+
+    if 1 != sum(x is not None for x in (success_probs, success_logits, failure_probs, failure_logits)):
+        raise ValueError("Exactly one of {success_probs, success_logits, failure_probs, failure_logits} must be specified")
+    ...math....
```

</details>

2. In a subsequent PR and for the 1.8 release, drop the `probs` and `logits` kwargs:

<details>

```diff
- def __init__(total_count, probs=None, logits=None, *,
+ def __init__(total_count, *,
               success_probs=None, success_logits=None,
               failure_probs=None, failure_logits=None,
               validate_args=None):
```
Alternatively we could add `success_probs` as the sole optional positional arg (though this is more dangerous)
```diff
- def __init__(total_count, probs=None, logits=None, *,
-              success_probs=None, success_logits=None,
+ def __init__(total_count, success_probs=None, *,
+             success_logits=None,
               failure_probs=None, failure_logits=None,
               validate_args=None):
```

</details>

After step 2, the positional behavior will be as in numpy, and we older code will be easy to adapt via kwargs.

**Comment by SplitInfinity** [comment]
Sounds like a plan! We can land my docs PR for now since the current docs are incorrect and update the docs as we go through the two steps.

**Comment by vitkl** [comment]
I am wondering whether code changes discussed here were introduced

@fritzo @jamestwebber I am also interested in `GammaPoisson`. I think it is a much more intuitive parameterization for single-cell genomics (https://www.nature.com/articles/s41588-021-00873-4).

**Comment by jamestwebber** [comment]
Well you're already using Pyro, so you're [all set](http://docs.pyro.ai/en/stable/distributions.html#gammapoisson)!

**Comment by vitkl** [comment]
This is also true. But I did not know GammaPoisson existed in pyro until
reading this issue.
